lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -666,8 +702,107 @@ class BuildOpGroups : public FuncOpPartialLoweringPattern {
     return createGroup<TGroupOp>(rewriter, getComponentState().getComponentOp(),
                                  block->front().getLoc(), groupName);
   }
+
+  /// Creates assignments within the provided group to the address ports of the
+  /// memoryOp based on the provided addressValues.
+  void assignAddressPorts(PatternRewriter &rewriter, Location loc,
+                          calyx::GroupInterface group, calyx::MemoryOp memoryOp,
+                          Operation::operand_range addressValues) const {
+    IRRewriter::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPointToEnd(group.getBody());
+    auto addrPorts = memoryOp.addrPorts();
+    assert(addrPorts.size() == addressValues.size() &&
+           "Mismatch between number of address ports of the provided memory "
+           "and address assignment values");
+    for (auto &idx : enumerate(addressValues))
+      rewriter.create<calyx::AssignOp>(loc, addrPorts[idx.index()], idx.value(),
+                                       Value());
+  }
 };
 
+LogicalResult BuildOpGroups::buildOp(PatternRewriter &rewriter,
+                                     memref::LoadOp loadOp) const {
+  auto memoryOp = getComponentState().getMemory(loadOp.memref());
+  if (singleLoadFromMemory(loadOp)) {
+    /// Single load from memory; Combinational case - we do not have to consider
+    /// adding registers in front of the memory.
+    auto combGroup = createGroupForOp<calyx::CombGroupOp>(rewriter, loadOp);
+    assignAddressPorts(rewriter, loadOp.getLoc(), combGroup, memoryOp,
+                       loadOp.getIndices());
+
+    /// We refrain from replacing the loadOp result with memoryOp.readData,
+    /// since multiple loadOp's need to be converted to a single memory's
+    /// ReadData. If this replacement is done now, we lose the link between
+    /// which SSA memref::LoadOp values map to which groups for loading a value
+    /// from the Calyx memory. At this point of lowering, we keep the
+    /// memref::LoadOp SSA value, and do value replacement _after_ control has
+    /// been generated (see LateSSAReplacement). This is *vital* for things such
+    /// as InlineCombGroups to be able to properly track which memory assignment
+    /// groups belong to which accesses.
+    getComponentState().registerEvaluatingGroup(loadOp.getResult(), combGroup);
+  } else {
+    auto group = createGroupForOp<calyx::GroupOp>(rewriter, loadOp);
+    assignAddressPorts(rewriter, loadOp.getLoc(), group, memoryOp,
+                       loadOp.getIndices());
+
+    /// Multiple loads from the same memory; In this case, we _may_ have a
+    /// structural hazard in the design we generate. To get around this, we
```

Comment:
Can you explain why this is a structural hazard? There's probably something obvious I'm missing here.

Reply:
If special care is not taken here, the combinational inliner could inline multiple assignment groups to the same memory. This i would see as a structural hazard; we're trying to a single-use resource at multiple locations in the same clock cycle.

Reply:
Makes sense. Thanks.